plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java

```diff
@@ -664,4 +670,117 @@ public boolean isAllowed(String groovyResourceUrl) {
         assertThat(logger.getMessages(), containsInRelativeOrder("finished suspending all executions", "ensuring all executions are saved"));
     }
 
+    @Issue("JENKINS-71692")
+    @Test public void stepsAreStoppedWhenCpsVmCroaks() throws Throwable {
+        sessions.then(r -> {
+            WorkflowJob p = r.createProject(WorkflowJob.class, "p");
+            p.setDefinition(new CpsFlowDefinition(
+                "parallel(\n" +
+                "  'willCroak': {\n" +
+                "    node {\n" +
+                "      label: {\n" + // Defines a label on a BlockStatement, which is erroneously ignored by CpsTransformer.
+                "        while (true) {\n" +
+                "          semaphore('croak')\n" +
+                "          break label\n" + // CallEnv.getBreakAddress throws an IllegalStateException when this runs.
+                "        }\n" +
+                "      }\n" +
+                "    }\n" +
+                "  },\n" +
+                "  'keepsRunning': {\n" +
+                "    semaphore('other')\n" +
+                "    echo 'kept running!'\n" +
+                "  }\n" +
+                ")", true));
+            WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("croak/1", b);
+            SemaphoreStep.waitForStart("other/1", b);
+            SemaphoreStep.success("croak/1", null);
+            r.assertBuildStatus(Result.FAILURE, r.waitForCompletion(b));
+            r.assertLogContains("Terminating parallel (id: 3)", b);
+            r.assertLogContains("Terminating node (id: 7)", b);
+            r.assertLogContains("Terminating semaphore (id: 8)", b);
+            for (Executor executor : Jenkins.get().toComputer().getExecutors()) {
+                // Node step should have cleaned up its PlaceholderExecutable.
+                assertThat(executor.getCurrentWorkUnit(), nullValue());
+            }
+            // Simulate some async step completing after CpsFlowExecution.croak.
+            SemaphoreStep.success("other/1", null);
+            // It's difficult to add meaningful non-flaky test assertions here.
+            // Even before the associated fix, 'kept running' wasn't printed, because although CpsThreadGroup.run
+            // did execute again, it hit an IllegalStateException in SandboxContinuable, which resulted in a second
+            // call to CpsFlowExecution.croak.
+            assertTrue(((CpsFlowExecution) b.getExecution()).programPromise.get(1, TimeUnit.SECONDS).runner.isShutdown());
+        });
+    }
+
+    @Issue("JENKINS-71692")
+    @Test public void stepsAreStoppedWhenBodyExecutionCallbackThrows() throws Throwable {
```

Comment:
Mimics the case described in [this comment](https://issues.jenkins.io/browse/JENKINS-70267?focusedId=434633&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-434633), although the callback was also fixed upstream, see https://github.com/jenkinsci/gitlab-plugin/pull/1410.

Reply:
Yes I also tested changes along those lines, but to start with I just wanted to show that the new fix is somewhat generic and covers various kinds of fatal VM thread errors. I will add the special case now though.

Reply:
Updated to match `TailCall` behavior in [a27f32f](https://github.com/jenkinsci/workflow-cps-plugin/pull/780/commits/a27f32fd7a35f3e847e10a429cc7c15c723df26f).